Efficient burn in for sampling

vgnecula opened the pull request:
Modified the sampling.hpp to solve:  Inefficient handling of burn in phase in sampling functions #180.

Handling burn in phase without storing the points required for this phase, for each of the sampling methods presented.

Cleared my branch non-related developments in order to emphasize just the sampling.hpp code.

TolisChal commented on include/sampling/sampling.hpp:
```diff
@@ -21,45 +21,37 @@
 
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, walk_len, rng); 
+        randPoints.push_back(p);  
```
Could you please reserve memory equal to rnum points before the push_backs? (i.e., randPoints.reserve(rnum);)

vgnecula replied:
I am not able to do that directly using randPoints.reserve(rnum); as randPoints is an instance of an std::list and not std::vector. Thus it does not support this type of command. Am I missing something?

vgnecula replied:
I thought about finding a workaround for reserving space from the first time, however as I figured out randPoints was a list, I decided that it is not the case anymore, as I can't get continuous storage anyway.

vfisikop replied:
`randPoints` is templated (i.e. it can be either a std::vector or a std::list or any container that supports `push_back`), thus you can create a specialized version of this function that reserves memory only for the case of a vector but I propose not to do anything in this PR. We should think if we need to want to keep it as it is or restrict `randPoints` to std::list or std::vector.
@TolisChal do we need random access in samples? If not should we restrict to std::list here?

TolisChal commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
+
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, walk_len, rng);  
+        randPoints.push_back(p);  
```
same as before

TolisChal commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
 
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, a, walk_len, rng); 
+        randPoints.push_back(p); 
```
same as before

TolisChal commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
+    
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, a, walk_len, rng); 
+        randPoints.push_back(p); 
```
same as before

TolisChal commented on include/sampling/sampling.hpp:
```diff
@@ -239,65 +219,40 @@ void logconcave_sampling(PointList &randPoints,
+            walk.apply(P, p, walk_len, rng); 
+        }
     }
-    logconcave_walk.disable_adaptive();
```
Why did you delete this line?

TolisChal commented on include/sampling/sampling.hpp:
```diff
@@ -309,54 +264,38 @@ void crhmc_sampling(PointList &randPoints,
-                  NegativeGradientFunctor
-          > walk_params;
-  Point p = Point(problem.center);
-  problem.options.simdLen=simdLen;
```
This line it is deleted and not re-written. Please be sure that your changes modify only how volesti stores the points and don't affect anything else.

TolisChal commented:
@vgnecula thanks for this PR!

A general comment is, when you want to change a particular part of the code, just change this one without affecting anything else. Then, it's easier to review the code and be sure about the changes.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
+        }
+    }
+
+    randPoints.clear();
```
clear is not needed here.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
-        typename RandomNumberGenerator,
-        typename Point
+
+template <
```
I would prefer to change style at a separate PR.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
-                                        randPoints, push_back_policy, rng);
 
+    
+    randPoints.clear();
```
is this needed?

vfisikop replied:
similar in all walks

vgnecula replied:
Initially, I did not have this function. However, I thought that it might be a good practice to ensure that everything is cleared before performing the actual sampling. Besides, this operation should be done in constant time (correct me if I am wrong), especially in the case in which we expect randPoints to be empty. Consequently, I thought that there would be no problem in adding it just for safety.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
+    
+    randPoints.clear();
+
+    for (unsigned int i = 0; i < rnum; ++i) {
```
there is a mistake here, it was rnum/2 before

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
+
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p1, p2, walk_len, rng); 
+        randPoints.push_back(p1); 
```
should we also push p2?

vgnecula replied:
From what I understood regarding how boundary sampling works, the p2 is a kind of dummy point used for processing the walk and updating p1. Thus, I think that we only need to push p1.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -68,169 +60,157 @@ template <
+    typename WalkTypePolicy::template Walk<Polytope, RandomNumberGenerator> walk(P, starting_point, a, rng);
 
-    Point p = starting_point;
+    Point p = starting_point; 
```
please do not add trailing whitespaces (there are more cases like this in this PR)

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -239,65 +219,40 @@ void logconcave_sampling(PointList &randPoints,
-                    NegativeGradientFunctor
-            > walk_params;
-
-    // Initialize random walk parameters
```
there is probably some mistake deleting this

vgnecula replied:
I do not use walk_params anymore.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -309,54 +264,38 @@ void crhmc_sampling(PointList &randPoints,
+        randPoints.push_back(crhmc_walk.getPoint()); 
+    }
 }
+
```
too many extra lines

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -434,169 +373,162 @@ crhmc_sampling <
+    const NT &a,
+    Point &starting_point, 
+    unsigned int const& nburns
+) {
```
looks like a new style of code, I suggest using the style of most of the functions in volesti

vfisikop reviewed (commented):
Thanks. I suggest as @TolisChal also suggested to focus on the issue of the burn in phase and leave style changes for future PR.

vfisikop commented:
Also it seems that you are deleting (by mistake) some crucial parts of the code in sampling.hpp.

vfisikop commented:
Side comment: there is no need to open a new PR every time you want to change something in your branch, just update your branch!

vgnecula commented:
@TolisChal @vfisikop just commited a new version, which adresses all your comments (find it as - minor probem with CRHMC solved - the last version of my new modifications). On my end it passes all tests. However, it takes way too much time to process crhmc_test_polytope_sampling. Besides, as you will probably see, I modified some functions, as I did not like some details that I implemented before.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -22,215 +22,193 @@
+                      RandomNumberGenerator &rng,
+                      const unsigned int &walk_len,
+                      const unsigned int &rnum,
+                      Point &starting_point,  
```
please avoid using trailing whitespaces.

vfisikop replied:
there are similar cases below

vgnecula replied:
Thank you for taking the time to review my new Commits!

I just wanted to be sure I understand what you mean by style changes.

So, I should have a first commit that for the lines where I did not change anything regarding the algorithm logic - the lines remain the same, while the second commit adds my proposed style changes.

I saw that there are some small differences in the ways some things are written and I tried to bring them all to the same writing style (that is why I made those modifications). Sorry for the white spaces, I missed some of them.

vfisikop replied:
Yes, exactly, we want one commit to fix the burn in phase that does not change any coding style (i.e. indentation etc) and another commit that contains all coding styles. Also please rebase (see `git rebase` documentation) your branch to only have those two commits and not the current ones (currently 4).

vgnecula replied:
@vfisikop should I push the "style-changes" commit? Or should I wait until the current one gets reviewed? Thank you!

vfisikop replied:
The PR should consist of two commits, one with the fixes and one with the style changes. So those could be reviewed independently.

vfisikop replied:
Having a look at this code again I see there are some problems with this interface, e.g. there is an issue of what walk should the user use for a particular distribution. You can have a look at this draft PR https://github.com/GeomScale/volesti/pull/300 for an idea of changing this to a new interface.

vgnecula replied:
I just started to look into it more. Could you please elaborate a bit more on what you think I should do? I want to be sure I fully understood what you meant.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -22,215 +22,193 @@
-                                            randPoints, push_back_policy, rng);
-        randPoints.clear();
+    for (unsigned int i = 0; i < nburns; ++i) {
+        walk.apply(P, current_point, dummy_point, walk_len, rng);
```
I think there is not reason to keep only one point, they are both on the boundary, so I propose to keep both. @TolisChal what do you think?

TolisChal replied:
Yes, we could do that, however keeping only one would be closer to a uniform sample (just intuitively).
A general comment from my side is that we need to delete this method in the future and implement Shake and Bake.

vfisikop commented on include/sampling/sampling.hpp:
```diff
@@ -239,65 +217,48 @@ void logconcave_sampling(PointList &randPoints,
-
-    // Initialize random walk parameters
-    unsigned int dim = starting_point.dimension();
-    walk_params params(F, dim);
```
how you pass the parameters to the walk?

vgnecula replied:
@vfisikop Thank you for pointing this out! So previously I was only relying on the assumption that the walk object will auto set the params. However after looking into how the langevin_walk is implemented, I saw that this is not the case. Thus, now I am setting the params using the set_params method. Let me know if this makes sense.

vfisikop reviewed (commented):
I reviewed part of it, but it is difficult to continue. Could you please separate your commits such that there is one commit with the changes to burn in feasibility and another one with the style changes?

vgnecula committed:
BurnIn Handling, same style
